### ISSUE TITLE
Bug in ERC20 contract, transfers can be disabled

### ISSUE THREAD

**Issue by VexyCats** [icon-foundation/ico#3]
In contract https://etherscan.io/address/0xb5a5f22694352c15b00323844ad545abb2b11028

There is a problem that allows anyone to call 

>     function disableTokenTransfer()
>     external
>     onlyFromWallet {
>         tokenTransfer = false;
>         TokenTransfer();
>     }

Because 

>  modifier onlyFromWallet {
>         require(msg.sender != walletAddress);
>         _;
>     }

token transfers are currently disabled.

**Comment by MeoMix** [comment]
Wow! How on earth did this make it to production? Where are your unit tests asserting these sort of conditions? If you don't have test coverage for this -- why are you handling millions of dollars and why should we trust you to handle money in the future?

**Comment by bitcartel** [comment]
It's not the first time a bug has been found in a token contract. So what's considered best practice these days to resolve a situation like this?  Fix bug, publish contract, initialized with clone of existing balances, communicate new address to ecosystem?

**Comment by MeoMix** [comment]
@bitcartel Yes, pretty much that. It's also possible to build contracts such that they're able to be upgraded, but this would've needed to be a design decision from the start. See https://consensys.github.io/smart-contract-best-practices/software_engineering/#upgrading-broken-contracts for more info

**Comment by postables** [comment]
It's also possible to write contracts, and pay more than a moderate amount of attention to the auditing process, and catch these bugs before they are released to production. ICX is worth what, 800 Million USD? And they couldn't be bothered to catch what is really something that shouldn't even make it past testing.

Rumour has it that it's the second time this very developer has made this very mistake:

https://twitter.com/minhokim/status/1007800274252062720?s=19
https://twitter.com/rgrpark/status/1007807240256962561

**Comment by dmdque** [comment]
The team can airdrop existing users a v2 token without much hassle. The hard part will be getting everyone to upgrade though.

It's insane that this bug has made it this far without being detected. But then again the ICX token contract wasn't even verified on Etherscan until a month or two ago so it didn't have many eyes on it.

**Comment by postables** [comment]
I proposed a fix in this reddit post

https://www.reddit.com/r/helloicon/comments/8rlgp0/workaround_for_icx_token_transfer_disable_bug/
